### PR TITLE
feat(ssl): Allow using .pem for public key file

### PR DESCRIPTION
**Pull request by randombk**
Try using $cert.pem as the certificate before falling back to
  $cert.crt, as .pem is the standard file extension for
  certificates. (RFC1421, RFC1424). This is different from the
  current behaviour, where only the .crt is used.

  NOTE: This could cause a regression if users have both a .pem and
  a .crt file in the /etc/nginx/certs directory. In that
  situation, the .pem will be used instead of the .crt, which
  can be problematic if the files contain different information
  (i.e. one contains the entire certificate chain, while the other
  does not). If this is an issue, the order can be changed so that
  the current behaviour is default.

Signed-off-by: David Li jiawei.davidli@gmail.com

### CODE REVIEW

**Comment by jwilder** [comment]
Would you be able to point me to the reference that says that `.pem` is the standard extension?

`.crt` seems pretty common.  https://en.wikipedia.org/wiki/X.509#Extensions_informing_a_specific_usage_of_a_certificate

**Comment by randombk** [comment]
Hmm. Looking closer, I see that the specification doesn't actually dictate a particular file extension to be used. However, I still believe we should default to `.pem`, for the following reasons:

* `.pem` files are unambiguously in PEM format, which is what Nginx requires. `.crt` files may be in either PEM or DER format.
* `.pem` is what the Nginx documentation uses in its examples [1]
* There is quite a lot of literature out there that tends to claim `.pem` as the default [2] [3] [4]

Thanks.

[1] http://nginx.org/en/docs/http/ngx_http_ssl_module.html#example

[2] https://www.digicert.com/ssl-certificate-installation-nginx.htm#edit_nginx

[3] https://support.ssl.com/Knowledgebase/Article/View/19/0/der-vs-crt-vs-cer-vs-pem-certificates-and-how-to-convert-them

[4] http://serverfault.com/a/9717

**Comment by md5** [comment]
I'm not sure how much @jwilder cares, but it looks like this PR does not maintain the existing whitespace. I know the files is currently a mix of hard tabs and spaces, but the lines being changed originally had tabs. I personally think the whole template should use tabs.

File: nginx.tmpl [hunk 1]
```diff
@@ -97,16 +102,17 @@ upstream {{ $host }} {
 {{ $certName := (first (groupByKeys $containers "Env.CERT_NAME")) }}
 
 {{/* Get the best matching cert  by name for the vhost. */}}
-{{ $vhostCert := (closest (dir "/etc/nginx/certs") (printf "%s.crt" $host))}}
+{{ $vhostCert := (closest (dir "/etc/nginx/certs") (printf "%s.key" $host))}}
```
Comment (md5) [inline nginx.tmpl]: This line seems weird. I think I understand why you changed it to look for `.key`, but it may need a comment.

Comment (randombk) [inline nginx.tmpl]: This entire vhost matching section is a bit off, tbh. I'll write up a separate issue/PR once I get the chance.

The three `{{ $vhostCert := replace $vhostCert ".[....] " "" -1 }}` lines below this look like they would break if the vhost name ends in .pem, .crt, or .key (for whatever reason). I.e. if the vhost name is `foo.key`, and the certificate file name is `foo.key.crt`, then the name could get stripped down to `foo`. I'll need to do a bit of testing to see if this is actually an issue, but I think the overall section needs a rework to cover these edge cases.

Comment (md5) [inline nginx.tmpl]: Luckily, `.crt`, `.key`, and `.pem` are not actual TLDs :) (see https://www.icann.org/resources/pages/listing-2012-02-25-en)

Comment (randombk) [inline nginx.tmpl]: Haha, true. Unfortunately, I know more than one network that uses those internally... It's definitely a rare edge case, but an edge case all the same.

**Comment by deini** [comment]
Any updates on this? I would also like to use `.pem`s
